pkg/eestream/aesgcm.go

```diff
@@ -0,0 +1,163 @@
+// Copyright (C) 2018 Storj Labs, Inc.
+// See LICENSE for copying information.
+
+package eestream
+
+import (
+	"crypto/aes"
+	"crypto/cipher"
+)
+
+type aesgcmEncrypter struct {
+	blockSize     int
+	key           [32]byte
+	startingNonce [12]byte
+	overhead      int
+	//aesgcm        AEAD
+}
+
+// NewAesGcmEncrypter returns a Transformer that encrypts the data passing
+// through with key.
+//
+// startingNonce is treated as a big-endian encoded unsigned
+// integer, and as blocks pass through, their block number and the starting
+// nonce is added together to come up with that block's nonce. Encrypting
+// different data with the same key and the same nonce is a huge security
+// issue. It's safe to always encode new data with a random key and random
+// startingNonce. The monotonically-increasing nonce (that rolls over) is to
+// protect against data reordering.
+//
+// When in doubt, generate a new key from crypto/rand and a startingNonce
+// from crypto/rand as often as possible.
+func NewAesGcmEncrypter(key *[32]byte, startingNonce *[12]byte,
+	encryptedBlockSize int) (Transformer, error) {
+	ptrkey := *key
+	slicekey := ptrkey[:]
+	block, err := aes.NewCipher(slicekey)
+	if err != nil {
+		panic(err.Error())
+	}
+	aesgcm, err := cipher.NewGCM(block)
+	if err != nil {
+		panic(err.Error())
+	}
+	if encryptedBlockSize <= aesgcm.Overhead() {
+		return nil, Error.New("block size too small")
+	}
+	return &aesgcmEncrypter{
+		blockSize:     encryptedBlockSize - aesgcm.Overhead(),
+		key:           *key,
+		startingNonce: *startingNonce,
+		overhead:      aesgcm.Overhead(),
+		//aesgcm:        aesgcmEncryp,
+	}, nil
+}
+
+func (s *aesgcmEncrypter) InBlockSize() int {
+	return s.blockSize
+}
+
+func (s *aesgcmEncrypter) OutBlockSize() int {
+	return s.blockSize + s.overhead
+}
+
+func calcGcmNonce(startingNonce *[12]byte, blockNum int64) (rv [12]byte,
+	err error) {
+	if copy(rv[:], (*startingNonce)[:]) != len(rv) {
+		return rv, Error.New("didn't copy memory?!")
+	}
+	_, err = incrementBytes(rv[:], blockNum)
+	return rv, err
+}
+
+func (s *aesgcmEncrypter) Transform(out, in []byte, blockNum int64) (
+	[]byte, error) {
+	key := s.key
+	slicekey := key[:]
+	block, err := aes.NewCipher(slicekey)
```

Comment:
You can just pass `s.key[:]` here instead of introducing two new local variables?

Reply:
modified as suggested